### PR TITLE
First implementation of migrations

### PR DESCRIPTION
**Pull request by tailhook**
Todos:
1. Interactive mode for `create-migration`
2. Tests (waiting for edgedb/edgedb#1654 to appear in nightly)

### CODE REVIEW

**Comment by tailhook** [comment]
Please, check error messages and other printed text.

File: src/migrations/create.rs [hunk 1]
```diff
@@ -1,14 +1,92 @@
...
+                    }
+                }
+                continue;
+            }
+            _ => return false,
+        }
+    }
+}
+
+#[context("error reading schema file {}", path.display())]
```
Comment (elprans) [inline src/migrations/create.rs:55]:
```suggestion
#[context("could not read schema file {}", path.display())]
```

File: src/migrations/create.rs [hunk 2]
```diff
@@ -1,14 +1,92 @@
...
+async fn read_schema_file(path: &Path) -> anyhow::Result<String> {
+    let data = fs::read_to_string(path).await?;
+    let mut offset = 0;
+    loop {
+        match full_statement(data[offset..].as_bytes(), None) {
+            Ok(shift) => offset += shift,
+            Err(_) => {
+                if !is_empty(&data[offset..]) {
+                    anyhow::bail!("Last statement must end with semicolon");
```
Comment (elprans) [inline src/migrations/create.rs:64]:
```suggestion
                    anyhow::bail!("final statement does not end with a semicolon");
```

File: src/migrations/create.rs [hunk 3]
```diff
@@ -1,14 +1,92 @@
...
+                }
+                return Ok(data);
+            }
+        }
+    }
+}
+
+#[context("error reading schema at {}", ctx.schema_dir.display())]
```
Comment (elprans) [inline src/migrations/create.rs:72]:
```suggestion
#[context("could not read schema in {}", ctx.schema_dir.display())]
```

File: src/migrations/status.rs [hunk 4]
```diff
@@ -1,10 +1,53 @@
...
+    -> Result<(), anyhow::Error>
+{
+    let mut items = cli.query::<CurrentMigration>(
+        "DESCRIBE CURRENT MIGRATION AS JSON",
+        &Value::empty_tuple(),
+    ).await?;
+    while let Some(data) = items.next().await.transpose()? {
+        if !data.confirmed.is_empty() || !data.proposed.is_empty() {
+            if !status.quiet {
+                eprintln!("Schema in database differ to on-disk schema, \
```
Comment (elprans) [inline src/migrations/status.rs]:
```suggestion
                eprintln!("Detected differences between the database schema and the schema source, \
```

File: src/migrations/migration.rs [hunk 5]
```diff
@@ -1,6 +1,398 @@
...
+                of the migration contents. To proceed you must fix the \
+                statement to read as:\n  \
+                CREATE MIGRATION {computed} ONTO ...\n
+                if this migration is not applied to \
+                any database or revert the changes to the file",
+                computed=id, file=migration.id);
+        }
+        Ok(())
+    } else {
+        anyhow::bail!("unknown version of migration id {:?}", migration.id);
```
Comment (elprans) [inline src/migrations/migration.rs:88]:
```suggestion
        anyhow::bail!("cannot parse migration name {:?}", migration.id);
```

File: src/migrations/migration.rs [hunk 6]
```diff
@@ -1,6 +1,398 @@
...
+    }
+}
+
+#[context("error reading migration file {}", path.display())]
```
Comment (elprans) [inline src/migrations/migration.rs:92]:
```suggestion
#[context("could not read migration file {}", path.display())]
```

File: src/migrations/migration.rs [hunk 7]
```diff
@@ -1,6 +1,398 @@
...
+        validate_text(&text, &data)?;
+    }
+    return Ok(data)
+}
+
+fn file_num(path: &Path) -> Option<u64> {
+    path.file_stem().and_then(|x| x.to_str()).and_then(|x| x.parse().ok())
+}
+
+#[context("error reading migrations in {}", dir.display())]
```
Comment (elprans) [inline src/migrations/migration.rs:108]:
```suggestion
#[context("could not read migrations in {}", dir.display())]
```

File: src/migrations/migration.rs [hunk 8]
```diff
@@ -1,6 +1,398 @@
...
+                v.insert(MigrationFile {
+                    path: path.to_path_buf(),
+                    data,
+                });
+            }
+            Entry::Occupied(o) => {
+                anyhow::bail!("Two files {:?} and {:?} have the same \
+                    parent revision {:?}. Multiple branches in revision \
+                    history are not supported yet, please rebase one of the \
+                    branches on top of other.",
```
Comment (elprans) [inline src/migrations/migration.rs:141]:
```suggestion
                    branches on top of the other.",
```

File: src/migrations/migration.rs [hunk 9]
```diff
@@ -1,6 +1,398 @@
...
+                        Some(n) => SortKey::Numeric(n),
+                        None => SortKey::Text(item.path.file_stem().unwrap()),
+                    }
+                })
+                .unwrap();
+            let valid_number = file_num(&next.path)
+                .map(|n| n == counter)
+                .unwrap_or(false);
+            if valid_number {
+                anyhow::bail!("File `{}` should have parent migration {:?} \
```
Comment (elprans) [inline src/migrations/migration.rs:177]:
```suggestion
                anyhow::bail!("File `{}` should have {:?} as the parent migration \
```

File: src/migrations/migration.rs [hunk 10]
```diff
@@ -1,6 +1,398 @@
...
+                    (`CREATE MIGRATION {} ONTO {} {{...`)",
+                    next.path.display(), parent_id,
+                    next.data.id, parent_id);
+            } else {
+                anyhow::bail!("Missing file `{:05}.edgeql` with \
```
Comment (elprans) [inline src/migrations/migration.rs:182]:
```suggestion
                anyhow::bail!("could not find file `{:05}.edgeql` with \
```

File: src/migrations/create.rs [hunk 11]
```diff
@@ -77,18 +92,116 @@ pub async fn gen_create_migration(ctx: &Context)
...
+pub async fn write_migration(ctx: &Context,
+    descr: &CurrentMigration, parent: &str, index: u64)
+    -> anyhow::Result<()>
+{
+    let dir = ctx.schema_dir.join("migrations");
+    let filename = dir.join(format!("{:05}.edgeql", index));
+    _write_migration(descr, parent, filename.as_ref()).await
+}
+
+#[context("error writing migration file {}", filepath.display())]
```
Comment (elprans) [inline src/migrations/create.rs:142]:
```suggestion
#[context("could not write migration file {}", filepath.display())]
```

**Pull request merged**